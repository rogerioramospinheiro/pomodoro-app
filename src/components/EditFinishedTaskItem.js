import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { editFinishedTask } from '../actions/tasks';

export class EditFinishedTaskItem extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            title: props.selectedTask.title,
            description: props.selectedTask.description
        };
    }
    onSubmitHandle = (e) => {
        e.preventDefault();
        const id = this.props.selectedTask.id;
        const updates = {
            title: this.state.title,
            description: this.state.description
        };
        this.props.editFinishedTask(id, updates);
        this.props.history.push('/');
    };
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState( {title} );
    };
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState( {description} );
    };
    render() {
        return (
            <div>
                <div className="timer-display">
                    <div className="timer-display__text">
                        <Link className="finished-tasks__link" to="/">Edit</Link>
                    </div>
                </div>
                <div className="task-form">
                    <form onSubmit={this.onSubmitHandle}>
                        <div className="task-form__spacebox">
                            <input type="text"
                            className="task-form__input"
                            placeholder="Title"
                            value={this.state.title}
                            onChange={this.onTitleChange}/>
                        </div>
                        <div className="task-form__spacebox">
                            <textarea
                            className="task-form__textarea"
                            placeholder="Description"
                            value={this.state.description}
                            onChange={this.onDescriptionChange}/>
                        </div>
                        <button className="task-form__save">Save</button>
                    </form>
                </div>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        selectedTask: state.tasks.finished_tasks.find( (task) => task.id === props.match.params.id )
    };
};

const mapDispatchToProps = (dispatch) => {
    return {
        editFinishedTask: (id, updates) => dispatch(editFinishedTask(id, updates))
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(EditFinishedTaskItem);