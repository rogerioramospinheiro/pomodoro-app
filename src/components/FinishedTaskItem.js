import React from 'react';
import moment from 'moment';
import { Link } from 'react-router-dom';

const FinishedTaskItem = ( {id, title, description, start_date, end_date} ) => {
    const start_moment = moment(start_date);
    const end_moment = moment(end_date);
    const duration = `${end_moment.diff(start_moment, 'minutes')}m`;
    return (
        <Link className="finished-tasks__link" to={`/edit/${id}`}>
            <div className="finished-tasks__item">
                <div>{ title }</div>
                <div>{ description }</div>
                <div>
                    <span className="finished-tasks__display-start-end">{ start_moment.format('hh:mm:ss A') }</span>
                    <span className="finished-tasks__display-start-end">-</span>
                    <span className="finished-tasks__display-start-end">{ end_moment.format('hh:mm:ss A') }</span>
                    <span className="finished-tasks__display-start-end">-</span>
                    <span>{ duration }</span>
                </div>
            </div>
        </Link>
    );
}

export default FinishedTaskItem;