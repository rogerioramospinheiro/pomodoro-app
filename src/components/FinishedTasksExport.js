import React from 'react';
import { CSVLink } from 'react-csv';
import moment from 'moment';

const FinishedTasksExport = ({tasks}) => {
    const headers = ['ID', 'TITLE', 'DESCRIPTION', 'START TIME', 'END TIME', 'DURATION'];
    const csvData = [headers];
    tasks.forEach( (task) => {
        const start_moment = moment(task.start_date);
        const end_moment = moment(task.end_date);
        const row = [ 
            task.id, 
            task.title, 
            task.description, 
            start_moment.format('hh:mm:ss A'), 
            end_moment.format('hh:mm:ss A'),
            `${end_moment.diff(start_moment, 'minutes')}m`
        ]
        csvData.push(row);
    });
    const printDate = moment(tasks[0].start_date).format('YYYYMMDD');
    const fileName = `pomodoro-app-export-${printDate}.csv`
    return (
        <CSVLink data={csvData} filename={fileName} enclosingCharacter={`"`} className="finished-tasks__link">
            <div className="task-form__button finished-tasks__link-export button-wide">Export</div>
        </CSVLink>
    );
};

export default FinishedTasksExport;