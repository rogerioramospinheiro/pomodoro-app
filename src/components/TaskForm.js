import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { startTask, cancelTask } from '../actions/tasks';

class TaskForm extends React.Component {
    state = {
        task_title: '',
        task_description: '',
    };
    onDescriptionChangeHandle = (e) => {
        const description = e.target.value;
        this.setState( {task_description: description} );
    };
    onTitleChangeHandle = (e) => {
        const title = e.target.value;
        this.setState( {task_title: title} );
    };
    toggleButtonName = () => {
        return this.isIdle() ? 'Start' : 'Stop';
    };
    onSubmitHandle = (e) => {
        e.preventDefault();
        if (this.isIdle()) {
            this.props.startTask({
                title: this.state.task_title, 
                description: this.state.task_description, 
                start_date: moment().valueOf(), 
            })
        } else {
            this.props.cancelTask();
        }
    };
    isIdle = () => {
        return this.props.state_name === 'IDLE';
    };
    isNotIdle = () => {
        return this.props.state_name !== 'IDLE';
    }
    render() {
        return (
            <div>
                <form onSubmit={this.onSubmitHandle}>
                    <input type="text"
                    placeholder="Task Title"
                    autoFocus
                    value={this.state.task_title}
                    onChange={this.onTitleChangeHandle}
                    disabled={this.isNotIdle()}
                    />
                    <input type="text"
                    placeholder="Task Description"
                    value={this.state.task_description}
                    onChange={this.onDescriptionChangeHandle}
                    disabled={this.isNotIdle()}
                    />
                    <button>{ this.toggleButtonName() }</button>
                </form>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        state_name: state.tasks.state_name
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        startTask: (task) => dispatch(startTask(task)),
        cancelTask: () => dispatch(cancelTask())
    }
};

export default connect(mapStateToProps, mapDispatchToProps)(TaskForm);