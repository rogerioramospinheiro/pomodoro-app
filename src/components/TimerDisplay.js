import React from 'react';
import { connect } from 'react-redux';
import { calculateTimer } from '../helpers/timerdisplay_helper';
import { finishTask } from '../actions/tasks';
import { selectInitTime, selectDisplayColor } from '../selectors/settings';

export class TimerDisplay extends React.Component {
    state = {
        minutes: 0,
        seconds: 0,
        endtime: 0
    };
    onIntervalHandle = (ctx) => {

        let minutes = ctx.state.minutes;
        let seconds = ctx.state.seconds;
        let end_time = ctx.state.endtime;

        if (ctx.props.init_time > 0) {

            const timer = calculateTimer(end_time, ctx.props.init_time);
            const updated_end_time = (timer.is_counting) ? timer.end_time : 0;

            ctx.setCounter(timer.minutes, timer.seconds, updated_end_time);

            if (timer.is_finished) {
                ctx.finishCountdown(end_time);
            }

        } else {

            if (minutes > 0 || seconds > 0) {
                ctx.setCounter(0, 0, 0);
            }

        }
    };
    setCounter = (minutes, seconds, endtime) => {
        this.setState( {minutes: minutes, seconds: seconds, endtime: endtime} );
    };
    displayCounter = () => {
        const m = this.state.minutes < 10 ? `0${this.state.minutes}` : this.state.minutes;
        const s = this.state.seconds < 10 ? `0${this.state.seconds}` : this.state.seconds;
        return `${m}:${s}`;
    };
    finishCountdown = (end_time) => {
        const now = {
            end_date: end_time
        }
        this.props.finishTask(now);
    };
    componentDidMount() {
        this.intervalTimeout = setInterval(this.onIntervalHandle, 500, this);
    }
    componentWillUnmount() {
        clearInterval(this.intervalTimeout);
    }
    render() {
        return (
            <div className="timer-display">
                <div className="timer-display__text" style={{color: this.props.display_color}}>{this.displayCounter()}</div>
            </div>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        init_time: selectInitTime(state.tasks, state.settings),
        display_color: selectDisplayColor(state.tasks)
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        finishTask: (end_date) => dispatch(finishTask(end_date))
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(TimerDisplay);