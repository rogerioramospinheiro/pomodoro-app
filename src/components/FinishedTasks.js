import React from 'react';
import { connect } from 'react-redux';
import FinishedTaskItem from './FinishedTaskItem';

export const FinishedTasks = (props) => (
    <div className="finished-tasks" >
        <FinishedTasksList {...props} />
    </div>
);

export const FinishedTasksList = (props) => {
    if (props.tasks.length === 0) {
        return (<div className="finished-tasks__item">No finished tasks</div>);
    } else {
        return props.tasks.map( forEachTask )
    }
}

const forEachTask = (task, index) => (
    <FinishedTaskItem key={index} {...task}/>
);

const mapStateToProps = (state) => {
    return {
        tasks: state.tasks.finished_tasks
    };
};

export default connect(mapStateToProps)(FinishedTasks);