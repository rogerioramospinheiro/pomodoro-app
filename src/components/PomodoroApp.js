import React from 'react';
import TaskForm from './TaskForm';
import TimerDisplay from './TimerDisplay';
import FinishedTasks from './FinishedTasks';

const PomodoroApp = () => (
    <div>
        <TimerDisplay />
        <TaskForm />
        <FinishedTasks />
    </div>
);

export default PomodoroApp;