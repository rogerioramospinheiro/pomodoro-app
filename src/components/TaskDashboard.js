import React from 'react';
import TaskForm from './TaskForm';
import TimerDisplay from './TimerDisplay';
import FinishedTasks from './FinishedTasks';

const TaskDashboard = () => (
    <div>
        <TimerDisplay />
        <TaskForm />
        <FinishedTasks />
    </div>
);

export default TaskDashboard;