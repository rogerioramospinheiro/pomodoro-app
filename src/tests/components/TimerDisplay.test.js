import React from 'react';
import { shallow } from 'enzyme';
import { TimerDisplay } from '../../components/TimerDisplay';

test('Should render TimerDisplay', () => {
    // Given
    const display_color ='black';
    const wrapper = shallow(<TimerDisplay display_color={display_color}/>);
    // Then
    expect(wrapper).toMatchSnapshot();
});