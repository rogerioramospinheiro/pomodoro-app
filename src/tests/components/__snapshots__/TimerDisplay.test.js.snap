// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render TimerDisplay 1`] = `
<div
  className="timer-display"
>
  <div
    className="timer-display__text"
    style={
      Object {
        "color": "black",
      }
    }
  >
    00:00
  </div>
</div>
`;
